Userland/ntpquery.cpp

```diff
@@ -0,0 +1,136 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define _GNU_SOURCE
+#include <LibCore/ArgsParser.h>
+#include <arpa/inet.h>
+#include <endian.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+
+struct [[gnu::packed]] NtpPacket {
+    uint8_t li_vn_mode;
+    uint8_t stratum;
+    int8_t poll;
+    int8_t precision;
+
+    uint32_t root_delay;
+    uint32_t root_dispersion;
+    uint32_t reference_id;
+
+    uint64_t reference_timestamp;
+    uint64_t origin_timestamp;
+    uint64_t receive_timestamp;
+    uint64_t transmit_timestamp;
+};
+
+static_assert(sizeof(NtpPacket) == 48);
+
+int main(int argc, char** argv)
+{
+    if (pledge("stdio inet dns", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    // FIXME: Request serenityos.pool.ntp.org here https://manage.ntppool.org/manage/vendor
+    // and then use that as the default value.
+    // Until then, explicitly pass this as `ntpquery pool.ntp.org`.
+    const char* host = nullptr;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(host, "NTP server", "host", Core::ArgsParser::Required::No);
+    args_parser.parse(argc, argv);
+
+    if (!host)
+        return 1;
+
+    auto* hostent = gethostbyname(host);
+    if (!hostent) {
+        printf("Lookup failed for '%s'\n", host);
+        return 1;
+    }
+
+    if (pledge("stdio inet", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+    unveil(nullptr, nullptr);
+
+    int fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+    if (fd < 0) {
+        perror("socket");
+        return 1;
+    }
+
+    struct timeval timeout {
+        5, 0
+    };
+    if (setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout)) < 0) {
+        perror("setsockopt");
+        return 1;
+    }
+
+    sockaddr_in peer_address;
+    memset(&peer_address, 0, sizeof(peer_address));
+    peer_address.sin_family = AF_INET;
+    peer_address.sin_port = htons(123);
+    peer_address.sin_addr.s_addr = *(const in_addr_t*)hostent->h_addr_list[0];
+
+    NtpPacket packet;
+    memset(&packet, 0, sizeof(packet));
+    packet.li_vn_mode = (4 << 3) | 3; // Version 4, client connection.
+
+    if (sendto(fd, &packet, sizeof(packet), 0, (const struct sockaddr*)&peer_address, sizeof(peer_address)) < 0) {
+        perror("sendto");
+        return 1;
+    }
+
+    socklen_t peer_address_size = sizeof(peer_address);
+    if (recvfrom(fd, &packet, sizeof(packet), 0, (struct sockaddr*)&peer_address, &peer_address_size) < 0) {
+        perror("recvfrom");
+        return 1;
+    }
+
```

Comment:
Don't we want to check here that the packet we've just recv'ed indeed comes from the peer we've sent the request to?
A connectionless protocol and all that, we can't be sure we're actually receiving a reply to the request, only that we've received a datagram sent to the same port...

Also, this should do a length check: `recvfrom()` returns a size and you have to compare that against `sizeof(packet)`

Reply:
Why not use a `Core::UDPServer`?
that'd have a nice secondary effect of making it easier to use SystemServer to spawn it...once we have IP socket takeover?

Reply:
Can `Core::UDPServer` be used with a preexisting socket?

Besides, if I understand @nico's intent here, it's meant to be a one-shot Userland utility for querying the time once, not a persistent service for timesync. And even if it was, I don't think socket takeover would be appropriate here — we're not listening for packets sent to us by clients (on a well-known port), we're waiting for a reply to our own request (on a random port).

Reply:
I doubt it can, but adding a ctor shouldn't be that difficult, I don't think?

> we're not listening for packets sent to us by clients (on a well-known port), we're waiting for a reply to our own request (on a random port).

Oh that's right; guess I wasn't thinking clearly there~

Reply:
The intent is to get something started. At the moment it's susceptible to MITM, but at the moment that's fine. If this one day becomes a system service, that needs fixing first then.

I looked at UDPSocket and it didn't do anything helpful. I expect this to grow a message loop and then that probably changes, but for now I don't see the point.

Reply:
Guess I should add a reply here too, for completeness sake.

IMO, checking that a function has initialized the whole buffer is not advanced functionality, it's basic C hygiene, much like checking return values for errors. You do all that even in simplest programs and tests (perhaps not in examples that are meant to showcase something, though, in order not to clutter the example).

Checking the peer address, in the other hand... I mean, it won't prevent MITM attacks either since we're not encrypting or signing anything. It just seems like a correctness thing to me, but whatever, I don't insist.

Reply:
I added a length check. I missed that part of your reply on first read, and this is a great catch indeed. Thanks for noticing this!